File: frontend/src/types/proposal.ts
export enum ProjectType {
  Website = 1,
  Ecommerce = 2,
  LandingPage = 3,
  System = 4,
  Mobile = 5,
  Integration = 6,
  Consulting = 7
}

export enum ProposalStatus {
  Sent = 1,
  UnderReview = 2,
  Accepted = 3,
  Refused = 4,
  Expired = 5
}

export interface Proposal {
  id: string;
  leadId: string;
  leadName: string;
  projectType: ProjectType;
  proposedValue: number;
  sentAt: string;
  status: ProposalStatus;
  refusalReason?: string | null;
  responseAt?: string | null;
  notes?: string | null;
}

export interface CreateProposalRequest {
  leadId: string;
  projectType: ProjectType;
  proposedValue: number;
  notes?: string;
}

export enum ProposalAction {
  MarkAsUnderReview = 0,
  Accept = 1,
  Refuse = 2,
  MarkAsExpired = 3,
  UpdateNotes = 4,
}

export interface UpdateProposalRequest {
  action: ProposalAction;
  refusalReason?: string;
  notes?: string;
}

File: backend/Codebymister.Domain/Entities/Proposal.cs
using Codebymister.Domain.Common;
using Codebymister.Domain.Enums;

namespace Codebymister.Domain.Entities;

public class Proposal : BaseEntity
{
    public Guid LeadId { get; private set; }
    public ProjectType ProjectType { get; private set; }
    public decimal ProposedValue { get; private set; }
    public DateTime SentAt { get; private set; }
    public ProposalStatus Status { get; private set; }
    public string? RefusalReason { get; private set; }
    public DateTime? ResponseAt { get; private set; }
    public string? Notes { get; private set; }

    public Lead Lead { get; private set; }

    private Proposal() { }

    public Proposal(
        Guid leadId,
        ProjectType projectType,
        decimal proposedValue,
        string? notes = null)
    {
        LeadId = leadId;
        ProjectType = projectType;
        ProposedValue = proposedValue;
        SentAt = DateTime.UtcNow;
        Status = ProposalStatus.Sent;
        Notes = notes;
    }

    public void MarkAsUnderReview()
    {
        Status = ProposalStatus.UnderReview;
    }

    public void Accept()
    {
        Status = ProposalStatus.Accepted;
        ResponseAt = DateTime.UtcNow;
    }

    public void Refuse(string reason)
    {
        Status = ProposalStatus.Refused;
        RefusalReason = reason;
        ResponseAt = DateTime.UtcNow;
    }

    public void MarkAsExpired()
    {
        Status = ProposalStatus.Expired;
    }

    public void UpdateNotes(string notes)
    {
        Notes = notes;
    }
}
